core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -262,6 +268,142 @@ public ShardIterator activeInitializingShardsIt(int seed) {
     }
 
     /**
+     * Returns an iterator over active and initializing shards, ordered by the adaptive replica
+     * selection forumla. Making sure though that its random within the active shards of the same
+     * (or missing) rank, and initializing shards are the last to iterate through.
+     */
+    public ShardIterator rankedActiveInitializingShardsIt(@Nullable ResponseCollectorService collector,
```

Comment:
down the road we really need to work on the number of objects being created in this process. It can totally be a followup but I think we can abstract it away quite nicely since it's all keyed by the node id and the set of nodes is static. We can use a bytesrefhash with parallel arrays in the future that also prevents all the boxing.

Reply:
nit: the other impl is called `activeInitializingShardsRandomIt` so maybe we should rename this one to `activeInitializingShardsRankedIt` or rename the other one to `randomActiveInitializingShardsIt`?